### ISSUE TITLE
C# Extension Thinks I have Visual Studio 2019 Installed

### ISSUE THREAD

**Issue by xobs** [dotnet/vscode-csharp#5159]
## Issue Description ##
When opening a project, Visual Studio Code prints the following information:

```
[info]: OmniSharp.Services.DotNetCliService
        Using the 'dotnet' on the PATH.
[info]: OmniSharp.Services.DotNetCliService
        DotNetPath set to dotnet
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 2 MSBuild instance(s)
            1: Visual Studio Community 2022 17.1.32328.378 17.1.0 - "C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin"
            2: StandAlone 17.1.1 - "c:\Users\Sean\.vscode\extensions\ms-dotnettools.csharp-1.24.3-win32-x64\.omnisharp\1.38.2\.msbuild\Current\Bin"
[warn]: OmniSharp.CompositionHostBuilder
        It looks like you have Visual Studio 2019 installed without .NET Core SDK support which is required by OmniSharp.
 Try updating Visual Studio 2019 installation with .NET Core SDK to enable better MSBuild support.
```

As you can see, the extension seems to think I have Visual Studio 2019 installed when in fact I do not. This is a fresh install of Windows 11, and has never had Visual Studio 2019 installed.

Furthermore, I installed .NET by going to the site opened by the extension: https://dotnet.microsoft.com/en-us/download/dotnet/thank-you/sdk-6.0.201-windows-x64-installer?journey=vs-code

The extension then refuses to work.

## Steps to Reproduce ##

1. Install Windows 11
2. Install Visual Studio 2022 Community
3. Install Visual Studio Code
4. Install C# extension
5. Open a directory with a `.csproj` file in it.

## Expected Behavior ##

Previously, the extension would open the project and allow me to auto-complete `.cs` files in the directory.

## Actual Behavior ##

The extension throws a lot of errors, such as:

```
C:\Program Files\dotnet\sdk\6.0.201\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(267,5): Error: The "ResolvePackageAssets" task failed unexpectedly.
NuGet.Packaging.Core.PackagingException: Unable to find fallback package folder 'C:\Program Files (x86)\Microsoft Visual Studio\Shared\NuGetPackages'.
```

### OmniSharp log ###
<details>Starting OmniSharp server at 08/04/2022, 12:09:33 pm
    Target: e:\Code\Xous\Core

OmniSharp server started.
    Path: c:\Users\Sean\.vscode\extensions\ms-dotnettools.csharp-1.24.3-win32-x64\.omnisharp\1.38.2\OmniSharp.exe
    PID: 20828

[info]: OmniSharp.Stdio.Host
        Starting OmniSharp on Windows 6.2.9200.0 (x64)
[info]: OmniSharp.Services.DotNetCliService
        Checking the 'DOTNET_ROOT' environment variable to find a .NET SDK
[info]: OmniSharp.Services.DotNetCliService
        Using the 'dotnet' on the PATH.
[info]: OmniSharp.Services.DotNetCliService
        DotNetPath set to dotnet
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 2 MSBuild instance(s)
            1: Visual Studio Community 2022 17.1.32328.378 17.1.0 - "C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Current\Bin"
            2: StandAlone 17.1.1 - "c:\Users\Sean\.vscode\extensions\ms-dotnettools.csharp-1.24.3-win32-x64\.omnisharp\1.38.2\.msbuild\Current\Bin"
[warn]: OmniSharp.CompositionHostBuilder
        It looks like you have Visual Studio 2019 installed without .NET Core SDK support which is required by OmniSharp.
 Try updating Visual Studio 2019 installation with .NET Core SDK to enable better MSBuild support.
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        MSBUILD_EXE_PATH environment variable set to 'c:\Users\Sean\.vscode\extensions\ms-dotnettools.csharp-1.24.3-win32-x64\.omnisharp\1.38.2\.msbuild\Current\Bin\MSBuild.exe'
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: StandAlone 17.1.1 - "c:\Users\Sean\.vscode\extensions\ms-dotnettools.csharp-1.24.3-win32-x64\.omnisharp\1.38.2\.msbuild\Current\Bin"
            CscToolExe = csc.exe
            CscToolPath = c:\Users\Sean\.vscode\extensions\ms-dotnettools.csharp-1.24.3-win32-x64\.omnisharp\1.38.2\.msbuild\Current\Bin\Roslyn
            MSBuildExtensionsPath = c:\Users\Sean\.vscode\extensions\ms-dotnettools.csharp-1.24.3-win32-x64\.omnisharp\1.38.2\.msbuild
            MSBuildToolsPath = c:\Users\Sean\.vscode\extensions\ms-dotnettools.csharp-1.24.3-win32-x64\.omnisharp\1.38.2\.msbuild\Current\Bin
[info]: OmniSharp.WorkspaceInitializer
        Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0
[info]: OmniSharp.Cake.CakeProjectSystem
        Detecting Cake files in 'e:\Code\Xous\Core'.
[info]: OmniSharp.Cake.CakeProjectSystem
        Did not find any Cake files
[info]: OmniSharp.MSBuild.ProjectSystem
        No solution files found in 'e:\Code\Xous\Core'
[info]: OmniSharp.MSBuild.ProjectManager
        Queue project update for 'e:\Code\Xous\Core\emulation\peripherals.csproj'
[info]: OmniSharp.Script.ScriptProjectSystem
        Detecting CSX files in 'e:\Code\Xous\Core'.
[info]: OmniSharp.MSBuild.ProjectManager
        Loading project: e:\Code\Xous\Core\emulation\peripherals.csproj
[info]: OmniSharp.Script.ScriptProjectSystem
        Did not find any CSX files
[info]: OmniSharp.WorkspaceInitializer
        Configuration finished.
[info]: OmniSharp.Stdio.Host
        Omnisharp server running using Stdio at location 'e:\Code\Xous\Core' on host 10916.
[fail]: OmniSharp.MSBuild.ProjectLoader
        The "ResolvePackageAssets" task failed unexpectedly.
NuGet.Packaging.Core.PackagingException: Unable to find fallback package folder 'C:\Program Files (x86)\Microsoft Visual Studio\Shared\NuGetPackages'.
   at NuGet.Packaging.FallbackPackagePathResolver..ctor(String userPackageFolder, IEnumerable`1 fallbackPackageFolders)
   at Microsoft.NET.Build.Tasks.NuGetPackageResolver.CreateResolver(IEnumerable`1 packageFolders)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheWriter..ctor(ResolvePackageAssets task)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheReader.CreateReaderFromDisk(ResolvePackageAssets task, Byte[] settingsHash)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheReader..ctor(ResolvePackageAssets task)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.ReadItemGroups()
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.ExecuteCore()
   at Microsoft.NET.Build.Tasks.TaskBase.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()
[warn]: OmniSharp.MSBuild.ProjectManager
        Failed to load project file 'e:\Code\Xous\Core\emulation\peripherals.csproj'.
e:\Code\Xous\Core\emulation\peripherals.csproj
C:\Program Files\dotnet\sdk\6.0.201\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(267,5): Error: The "ResolvePackageAssets" task failed unexpectedly.
NuGet.Packaging.Core.PackagingException: Unable to find fallback package folder 'C:\Program Files (x86)\Microsoft Visual Studio\Shared\NuGetPackages'.
   at NuGet.Packaging.FallbackPackagePathResolver..ctor(String userPackageFolder, IEnumerable`1 fallbackPackageFolders)
   at Microsoft.NET.Build.Tasks.NuGetPackageResolver.CreateResolver(IEnumerable`1 packageFolders)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheWriter..ctor(ResolvePackageAssets task)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheReader.CreateReaderFromDisk(ResolvePackageAssets task, Byte[] settingsHash)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.CacheReader..ctor(ResolvePackageAssets task)
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.ReadItemGroups()
   at Microsoft.NET.Build.Tasks.ResolvePackageAssets.ExecuteCore()
   at Microsoft.NET.Build.Tasks.TaskBase.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()

[fail]: OmniSharp.MSBuild.ProjectManager
        Attempted to update project that is not loaded: e:\Code\Xous\Core\emulation\peripherals.csproj
[warn]: OmniSharp.Roslyn.CSharp.Services.InlayHints.InlayHintService
        Inlay hints requested for document not in workspace Location { FileName = e:\Code\Xous\Core\emulation\peripherals\ComSoC.cs, Range = Range { Start = Point { Line = 0, Column = 0 }, End = Point { Line = 126, Column = 0 } } }
[warn]: OmniSharp.Roslyn.CSharp.Services.InlayHints.InlayHintService
        Inlay hints requested for document not in workspace Location { FileName = e:\Code\Xous\Core\emulation\peripherals\ComSoC.cs, Range = Range { Start = Point { Line = 0, Column = 0 }, End = Point { Line = 99, Column = 0 } } }
[warn]: OmniSharp.Roslyn.CSharp.Services.InlayHints.InlayHintService
        Inlay hints requested for document not in workspace Location { FileName = e:\Code\Xous\Core\emulation\peripherals\ComSoC.cs, Range = Range { Start = Point { Line = 0, Column = 0 }, End = Point { Line = 103, Column = 0 } } }
[warn]: OmniSharp.Roslyn.CSharp.Services.InlayHints.InlayHintService
        Inlay hints requested for document not in workspace Location { FileName = e:\Code\Xous\Core\emulation\peripherals\ComSoC.cs, Range = Range { Start = Point { Line = 0, Column = 0 }, End = Point { Line = 104, Column = 9 } } }
</details>

### C# log ###
The C# log is empty
<details></details>

## Environment information ##

**VSCode version**: 1.66.1
**C# Extension**: 1.24.3

<details><summary>Dotnet Information</summary>
.NET SDK (reflecting any global.json):
 Version:   6.0.201
 Commit:    ef40e6aa06

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.201\

Host (useful for support):
  Version: 6.0.3
  Commit:  c24d9a9c91

.NET SDKs installed:
  6.0.201 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.3 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.3 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 6.0.3 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
</details>
<details><summary>Visual Studio Code Extensions</summary>

|Extension|Author|Version|
|---|---|---|
|better-cpp-syntax|jeff-hykin|1.15.13|
|cmake|twxs|0.0.17|
|cmake-tools|ms-vscode|1.10.5|
|cpptools|ms-vscode|1.9.7|
|cpptools-extension-pack|ms-vscode|1.1.0|
|cpptools-themes|ms-vscode|1.0.0|
|csharp|ms-dotnettools|1.24.3|
|doxdocgen|cschlosser|1.4.0|
|jupyter|ms-toolsai|2022.3.1000901801|
|jupyter-keymap|ms-toolsai|1.0.0|
|jupyter-renderers|ms-toolsai|1.0.6|
|markdown-all-in-one|yzhang|3.4.0|
|python|ms-python|2022.4.1|
|remote-containers|ms-vscode-remote|0.231.4|
|remote-ssh|ms-vscode-remote|0.78.0|
|remote-ssh-edit|ms-vscode-remote|0.80.0|
|remote-wsl|ms-vscode-remote|0.66.0|
|vscode-pylance|ms-python|2022.4.0|;

</details>

**Comment by xobs** [comment]
Additionally, the logs tell me to `Try updating Visual Studio 2019 installation with .NET Core SDK to enable better MSBuild support.`, however there is no ".NET Core SDK" in the Visual Studio Community 2022 installer:

![image](https://user-images.githubusercontent.com/238325/162361920-82a19cf5-c2f4-4b7a-968b-ae2b16abfda3.png)

**Comment by xobs** [comment]
Part of the problem appears to be that the extension assumes the following directories exist, but they didn't by default:

```
C:\Program Files\dotnet\sdk\NuGetFallbackFolder
C:\Program Files (x86)\Microsoft Visual Studio\Shared\NuGetPackages
```

I created these directories without anything in them, and Omnisharp at least succeeded in starting up.

**Comment by filipw** [comment]
The error message is hardcoded and therefore does not reflect the state of your system accurately.
Between 2022 and 2019, the SDK was reamed to .NET SDK (not .NET Core SDK), which is the checkbox you need to tick.
<img width="1381" alt="Bildschirmfoto 2022-04-08 um 09 32 44" src="https://user-images.githubusercontent.com/1710369/162395141-3cc79941-27a1-45b1-8b4d-0d4a9bc05bb7.png">

Alternatively, you can tell OmniSharp to not discover MSBuild from your VS installation, and instead only use the one available with the .NET SDK. This can be achieved by setting

```
"omnisharp.useModernNet": true
```

This will use the .NET 6.0 build of OmniSharp and will not interact with your VS installation anymore. This option is currently opt-in, but will be made default in the next release.

**Comment by xobs** [comment]
Thanks for the input! I tried setting `omnisharp.useModernNet: true` and it didn't work. Creating the directories I mentioned above helped things somewhat.

So it seems that there are two issues that need addressing:

1. The version shouldn't be hardcoded, it should print the actual version that's installed.
2. The name should reflect the name that's present in the version that's installed.

**Comment by filipw** [comment]
> I tried setting `omnisharp.useModernNet: true` and it didn't work.

can you please post the output log that you get when this flag is enabled?

**Comment by filipw** [comment]
> Creating the directories I mentioned above helped things somewhat.

this error is really coming from the MSBuild in Visual Studio. If you installed the .NET SDK component in VS, there is no need to create them manually.

**Comment by xobs** [comment]
It looks like modern .NET doesn't work at all:

```
Starting OmniSharp server at 08/04/2022, 5:13:22 pm
    Target: e:\Code\Xous\Core

OmniSharp server started with .NET 6.0.201
.
    Path: c:\Users\Sean\.vscode\extensions\ms-dotnettools.csharp-1.24.3-win32-x64\.omnisharp\1.38.2-net6.0\OmniSharp.dll
    PID: 40372

Could not execute because the specified command or file was not found.
Possible reasons for this include:
  * You misspelled a built-in dotnet command.
  * You intended to execute a .NET program, but dotnet-c:\Users\Sean\.vscode\extensions\ms-dotnettools.csharp-1.24.3-win32-x64\.omnisharp\1.38.2-net6.0\OmniSharp.dll does not exist.
  * You intended to run a global tool, but a dotnet-prefixed executable with this name could not be found on the PATH.
```

Creating those directories seems to have fixed "non-modern .NET" to the point where things work now.

**Comment by filipw** [comment]
> It looks like modern .NET doesn't work at all:

Because it is currently opt-in, the .NET 6.0 build of OmniSharp is not packaged with the extension and instead acquired on demand. Once you set the flag in the config, uninstall and install the extension again, and restart VS Code to force the download of this dependency.

**Comment by xobs** [comment]
I see. Then there is a third issue in that it's not installed on-demand, nor is there any explanation about why that is. Are there issues already for these three issues?


1. The version shouldn't be hardcoded, it should print the actual version that's installed.
2. The name should reflect the name that's present in the version that's installed.
3. Setting `omnisharp.useModernNet: true` gives an unhelpful error

**Comment by filipw** [comment]
> Then there is a third issue in that it's not installed on-demand

It did not install when you reinstalled extension with the flag turned on?

**Comment by JoeRobich** [comment]
@xobs Can you share the C# Log from the Output pane? It logs the attempt to acquire the O# build.

**Comment by xobs** [comment]
It did install, after I removed the extension and re-added it. The issue is that when I set `omnisharp.useModernNet: true` I was given the option to restart VS Code, but `OmniSharp for Windows (.NET 6 / x64)` was not downloaded. Instead, the user needs to remove the entire extension and re-install it with the setting already enabled.

It was surprising behavior, but if it's expected then it doesn't warrant opening an issue. Particularly since that it will default to using that method in the future.
